### PR TITLE
9 upload doc to pas package

### PR DESCRIPTION
**Pull request by godfreyyeung**
This PR/commit acts as a reference to address #9 

This is the minimal set of modules, controllers and services to connect to CRM and make a document upload.

Most of it was taken from `labs-zap-api` and stripped of unecessary imports.

### CODE REVIEW

File: server/src/document/document.controller.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    } else if ( entityName === 'dcp_package' ) {
+      const packageGUID = instanceId;
+
+      const packageRecord = (await CRMWebAPI.get(`dcp_packages?$select=dcp_name&$filter=dcp_packageid eq '${packageGUID}'&$top=1`))['value'][0];
+
+      const packageName = packageRecord.dcp_name;
+
+      const folderName = `${packageName}_${packageGUID.replace(/\-/g,'').toUpperCase()}`;
+
+      uploadDocResponse = await CRMWebAPI.uploadDocument('dcp_artifacts', packageGUID, folderName, file.originalname, encodedBase64File, true, headers);
```
Comment (godfreyyeung) [inline server/src/document/document.controller.ts]: The first parameter to uploadDocument should be `dcp_packages`

**Comment by godfreyyeung** [comment]
> This is awesome! Would you mind adding a more in-depth description on the PR? Having trouble knowing where to start/would love an explanation of how this was implemented file-by-file etc. Even if it is the same as zap-api, I think it would help a lot in reviewing

can do!

File: server/src/app.module.ts [hunk 2]
```diff
@@ -1,10 +1,55 @@
...
 @Module({
-  imports: [],
+  imports: [
+    ConfigModule,
+    AuthModule,
+    DocumentModule,
+    TypeOrmModule.forRootAsync({
+      useFactory: (config: ConfigService) => ({
+        type: 'postgres',
+        url: config.get('DATABASE_URL'),
```
Comment (allthesignals) [inline server/src/app.module.ts:22]: This doesn't need to be in here. This is no database for this API.

File: server/src/config/config.module.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { Module } from '@nestjs/common';
+import { ConfigService } from './config.service';
+
+@Module({
+  providers: [
+    {
+      provide: ConfigService,
+      useValue: new ConfigService(`development.env`),
```
Comment (allthesignals) [inline server/src/config/config.module.ts:8]: Unnecessary template literal.

File: server/src/app.controller.ts [hunk 4]
```diff
@@ -1,12 +1,38 @@
...
+    } catch (e) {
+      if (e instanceof HttpException) {
+        res.status(401).send({ errors: [e] });
+      } else {
+        console.log(e);
+
+        res.status(500).send({ errors: [e] });
+      }
+    }
+  }
```
Comment (allthesignals) [inline server/src/app.controller.ts:36]: Why is this here? If this is for document upload.

File: server/src/auth.middleware.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+    try {
+      req.session = await this.authService.validateCurrentToken(token);
+
+      next();
+    } catch (e) {
+      proceedNoAuth(res, next);
+    }
+  }
+}
```
Comment (allthesignals) [inline server/src/auth.middleware.ts:26]: The PR says it is intended to address document upload (which was a research task, not really something to build out), but this is authentication logic. This seems unrelated to the issue at hand.

File: server/src/auth/auth.module.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    setTimeout(async () => {
+      console.log('generating cookie...');
+
+      const response = await superagent.get(`http://localhost:3000/login?accessToken=${mockJWT}`);
+      const { header: { 'set-cookie': [, token] } } = response;
+
+      console.log(`SKIP_AUTH is true! The cookie token is ${token}. Add this to your request headers.`);
+    }, 3000);
+  }
+}
```
Comment (allthesignals) [inline server/src/auth/auth.module.ts:54]: This was lifted from ZAP API to assist with development but is it really necessary for right now?

File: server/src/contact/contact.entity.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+  @Column()
+  fullname: string;
+
+  @Column()
+  statecode: string;
+
+  @Column()
+  statuscode: string;
+}
```
Comment (allthesignals) [inline server/src/contact/contact.entity.ts:19]: Again, this is a TypeORM object, which is specifically meant for modeling SQL database objects. There is no SQL database for this project. We wouldn't even need to use this.

File: server/src/contact/contact.service.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  async findOne(opts: any): Promise<Contact> {
+    return this.contactRepository.findOneOrFail(opts);
+  }
+
+  async findByEmail(email): Promise<Contact> {
+    return this.activeContacts()
+      .andWhere('emailaddress1 ILIKE :email', { email })
+      .getOne();
+  }
+}
```
Comment (allthesignals) [inline server/src/contact/contact.service.ts:29]: None of this would be necssary or woudl work because we aren't using a SQL data ase for this. THis is all TypeORM stuff which is specific to SQL datbases, not WEb API

File: server/src/document/document.controller.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      MSCRMCallerID: this.config.get('CRM_ADMIN_SERVICE_USER')
+    };
+
+    // encode base64
+    const encodedBase64File = Buffer.from(file.buffer).toString('base64');
+
+    let uploadDocResponse = {};
+
+    if ( entityName === 'dcp_package' ) {
+      const packageGUID = instanceId;
```
Comment (allthesignals) [inline server/src/document/document.controller.ts:59]: Why make another variable for this? If you need to alias it, you can do that in the destructuring above:

```
      body: {
        instanceId: packageGUID,
```

I think a comment would be fine, though.

File: server/src/document/document.controller.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    } = request;
+
+    const headers = {
+      MSCRMCallerID: this.config.get('CRM_ADMIN_SERVICE_USER')
+    };
+
+    // encode base64
+    const encodedBase64File = Buffer.from(file.buffer).toString('base64');
+
+    let uploadDocResponse = {};
```
Comment (allthesignals) [inline server/src/document/document.controller.ts:56]: Why is this defined up here? If the await fails below, it will throw a regular error anyway. And if needs to be a default object, that could be handled in a clearer way. I would just define a variable below

File: server/src/document/document.controller.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+    if ( entityName === 'dcp_package' ) {
+      const packageGUID = instanceId;
+
+      const packageRecord = (await CRMWebAPI.get(`dcp_packages?$select=dcp_name&$filter=dcp_packageid eq '${packageGUID}'&$top=1`))['value'][0];
+
+      const packageName = packageRecord.dcp_name;
```
Comment (allthesignals) [inline server/src/document/document.controller.ts:63]: You could destructure this:

`const { dcp_name: packageName } = (await ...`

File: server/src/document/document.controller.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+
+      const folderName = `${packageName}_${packageGUID.replace(/\-/g,'').toUpperCase()}`;
+
```
Comment (allthesignals) [inline server/src/document/document.controller.ts:66]: I would have put stuff like this in a separate function below `generateFolderName(dcp_name, instanceId)`

File: server/src/document/document.controller.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      uploadDocResponse = await CRMWebAPI.uploadDocument('dcp_package', packageGUID, folderName, file.originalname, encodedBase64File, true, headers);
```
Comment (allthesignals) [inline server/src/document/document.controller.ts:67]: Why not destructure attrs out of file, too? We do it for the request object above, and file is provided right there in the method signature.

**Comment by godfreyyeung** [comment]
No longer relevant to main stream of development work. Will pull out document research work for later feature branches. Thank you @allthesignals  for all the feedback!
